src/io.c

```diff
@@ -2075,7 +2075,9 @@ _dispatch_stream_handler(void *ctx)
 		_dispatch_stream_complete_operation(stream, op);
 		// Fall through
 	case DISPATCH_OP_RESUME:
-		if (_dispatch_stream_operation_avail(stream)) {
+		if (_dispatch_stream_operation_avail(stream) && stream->source) {
+			dispatch_async_f(stream->dq, stream->dq, _dispatch_stream_queue_handler);
```

Comment:
I have tested the change and it works as well as this one.

> this will likely cause a 100% CPU

It is not at all immediately obvious to me why that would be the case. I thought I was just checking if the source was not null. How would that lead to 100% CPU usage?

Reply:
you don't get what I mean, you should not do:
```diff
+			dispatch_async_f(stream->dq, stream->dq, _dispatch_stream_queue_handler);
```

this will cause a 100% CPU until the source fires.

You need to do:

```C
 	case DISPATCH_OP_RESUME:
 -		if (_dispatch_stream_operation_avail(stream)) {
 +		if (_dispatch_stream_operation_avail(stream) && !stream->source_running) {
  			stream->source_running = true;
  			dispatch_resume(_dispatch_stream_source(stream, op));
  		}
```

If `source_running` is true you're waiting for it to fire, and you should not do anything, I think

Reply:
So here is what I'll say:

1. It is still not clear to me how doing this `dispatch_async_f(stream->dq, stream->dq, _dispatch_stream_queue_handler)` will lead to 100% CPU usage. Are you positing that `read` would just return `EAGAIN` in an infinite loop thus leading to `_dispatch_stream_queue_handler` also executing in a tight infinite loop? It would seem to me that if `read` is always returning what is effectively a status code for "try again" then the issue would be with `read` in that case. But that might be naive on my part.
2. Your suggestion does not fix the bug described in [this comment of SR-5773](https://bugs.swift.org/browse/SR-5773?focusedCommentId=35547&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-35547). When I switch to your patch the bug returns. You can look through the dispatch logs, e.g., [run-002.log](https://github.com/apple/swift-corelibs-libdispatch/files/2063922/run-002.log), for a failing case with the patch below applied. Notice that you get the `EAGAIN` and then call through to `_dispatch_stream_source` then the "hang" on line 1974 then the infinite loop begins.

```patch
diff --git a/src/io.c b/src/io.c
index 33a87fe..5990446 100644
--- a/src/io.c
+++ b/src/io.c
@@ -1971,6 +1971,7 @@ _dispatch_stream_source(dispatch_stream_t stream, dispatch_operation_t op)
 {
 	// On stream queue
 	if (stream->source) {
+		_dispatch_op_debug("hang", op);
 		return stream->source;
 	}
 	dispatch_fd_t fd = op->fd_entry->fd;
@@ -2075,9 +2076,7 @@ pick:
 		_dispatch_stream_complete_operation(stream, op);
 		// Fall through
 	case DISPATCH_OP_RESUME:
-		if (_dispatch_stream_operation_avail(stream) && stream->source) {
-			dispatch_async_f(stream->dq, stream->dq, _dispatch_stream_queue_handler);
-		} else if (_dispatch_stream_operation_avail(stream)) {
+		if (_dispatch_stream_operation_avail(stream) && !stream->source_running) {
 			stream->source_running = true;
 			dispatch_resume(_dispatch_stream_source(stream, op));
 		}
```



The whole point of calling `_dispatch_stream_source` is to:

1. Create a [`stream`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L1978-L1988)
2. Register [`_dispatch_stream_source_handler`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L1989-L1991) for the first time
3. Which in turn starts the calls to [`_dispatch_stream_handler`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L2006-L2010)

My patch just short circuits the loop that would have been created had `_dispatch_stream_source` not returned early, similar to [`DISPATCH_OP_COMPLETE`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L2069-L2072), and elides steps 1 and 2 above.

Maybe the solution is just to call clear the event handler and call [`dispatch_source_set_event_handler_f`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L1989-L1991) in the [early return](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L1973-L1975)? That seems like it is riskier because I don't fully understand all the places and conditions under which `_dispatch_stream_source` is called but `DISPATCH_OP_RESUME` seems like a targeted condition that is specifically for [`EAGAIN`](https://github.com/apple/swift-corelibs-libdispatch/blob/swift-4.2-branch/src/io.c#L2432-L2441).

Or maybe something else entirely? I am open to other suggestions. :man_shrugging:

Reply:
`_dispatch_stream_queue_handler()` calls `_dispatch_stream_handler()` which calls `_dispatch_operation_perform()` and `read()` which returns `EAGAIN` again, and will do so until the read() result changes.

this is invalid.

But if the source is armed, then when the EAGAIN -> not EAGAIN flips, `_dispatch_stream_source_handler()` will run which runs `_dispatch_stream_handler()` again and kicks things back

WHich is exactly why in that codepath, you should check for `if the source is running`, if yes, you just have to wait for it to fire, if not then you're good.

If my patch doesn't work, then there's something off with the dispatch source not firing, or not being resumed properly on Linux, not with io.